daemon/daemon.go

```diff
@@ -1139,7 +1139,17 @@ func NewDaemon() (*Daemon, error) {
 
 	// Set up ipam conf after init() because we might be running d.conf.KVStoreIPv4Registration
 	log.Info("Initializing IPAM")
-	switch err := ipam.Init(); err.(type) {
+	ipam.Init()
+
+	// restore endpoints before any IPs are allocated to avoid eventual IP
+	// conflicts later on, otherwise any IP conflict will result in the
+	// endpoint not being able to be restored.
+	restoredEndpoints, err := d.restoreOldEndpoints(option.Config.StateDir, true)
```

Comment:
Doesn't this need to be gated behind an `if option.Config.RestoreState {` like `regenerateRestoredEndpoints` below? Or, perhaps, we always "restore" them but we would clobber/delete them  (with `d.deleteEndpointQuiet` as we do on error inside `restoreOldEndpoints`). Otherwise we have these endpoints but we never regenerate them when restore=false.

Reply:
restoreOldEndpoints() returns an empty list if restoring is disabled.

I agree that we should remove old state when restore is disabled. We should do it in a separate PR though. This is already getting big.

Reply:
Ah, indeed `restoreOldEndpoints` does. I would prefer not to do that but I don't want to debate global variables again.